File: 01/cpp/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <cstdlib>

#define ABS(X)	(((X) < 0) ? -1 * (X) : (X))

static bool cmp(int a, int b) {
	return a < b;
}

bool ReadInts(std::string path, std::vector<int>& left, std::vector<int>& right) {
	std::ifstream in(path, std::ios_base::in);
	std::string line, word;
	while (getline(in, line)) {
		std::stringstream ss(line);
		ss >> word;
		left.push_back(atoi(word.c_str()));
		ss >> word;
		right.push_back(atoi(word.c_str()));
	}
	std::sort(left.begin(), left.end(), cmp);
	std::sort(right.begin(), right.end(), cmp);
	return true;
}

int Part1(std::vector<int>& left, std::vector<int>& right) {
	int total = 0;
	if (left.size() != right.size()) {
		std::cout << "Lengths are not the same" << std::endl;
		exit(-1);
	}
	for (size_t i = 0; i < left.size(); i++) {
		total += ABS(left[i] - right[i]);
	}
	return total;
}

int Part2(std::vector<int>& left, std::vector<int>& right) {
	int total = 0;
	if (left.size() != right.size()) {
		std::cout << "Lengths are not the same" << std::endl;
		exit(-1);
	}
	std::vector<int> counts;
	counts.resize(*right.end() + 1);
	for (auto i = right.begin(); i != right.end(); i++)
		counts[*i]++;
	for (auto i = left.begin(); i != left.end(); i++) {
		total += *i * counts[*i];
	}
	return total;
}

int main(int argc, char *argv[]) {
	if (--argc != 2) {
		std::cout << "Usage: ./main [1|2] file" << std::endl;
		return -1;
	}
	std::string path = argv[2];
	bool part = strcmp(argv[1], "1") == 0;
	std::vector<int> left, right;
	if (!ReadInts(path, left, right))
		return -1;
	int score;
	if (part) {
		score = Part1(left, right);
		std::cout << "Total Difference: " << score << std::endl;
	} else {
		score = Part2(left, right);
		std::cout << "Similarity Score: " << score << std::endl;
	}
	return 0;
}

File: 01/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func ReadInts(path string) ([]int, []int) {
	input, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer input.Close()
	left := make([]int, 0)
	right := make([]int, 0)
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		nums := strings.Split(scanner.Text(), "   ")
		num1, err := strconv.Atoi(nums[0])
		if err != nil {
			panic(err)
		}
		num2, err := strconv.Atoi(nums[1])
		if err != nil {
			panic(err)
		}
		left = append(left, num1)
		right = append(right, num2)
	}
	sort.Slice(left, func(a, b int) bool {
		return left[a] < left[b]
	})
	sort.Slice(right, func(a, b int) bool {
		return right[a] < right[b]
	})
	return left, right
}

func Part1(path string) int {
	left, right := ReadInts(path)
	if len(left) != len(right) {
		panic("Different Length Slices")
	}
	var total int = 0
	for i, l := range left {
		total += int(math.Abs(float64(l - right[i])))
	}
	return total
}

func Part2(path string) int {
	left, right := ReadInts(path)
	if len(left) != len(right) {
		panic("Different Length Slices")
	}
	totals := make([]int, right[len(right)-1]+1)
	for _, r := range right {
		totals[r] += 1
	}
	total := 0
	for _, l := range left {
		total += l * totals[l]
	}
	return total
}

func main() {
	if len(os.Args) != 3 {
		panic("Usage: ./main [1|2] file")
	}
	switch os.Args[1] {
	default:
	case "1":
		fmt.Println("Total Distance:", Part1(os.Args[2]))
	case "2":
		fmt.Println("Similarity Score:", Part2(os.Args[2]))
	}
}
